File: app/src/main/java/com/asmat/rolando/popularmovies/database/MoviesDAO.java
package com.asmat.rolando.popularmovies.database;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.*;

@Dao
public interface MoviesDAO {

    // Movie

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertMovie(Movie movie);

    @Delete
    void deleteMovie(Movie movie);

    // Favorite

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertFavoriteMovie(FavoriteMovie movie);

    @Delete
    void deleteFavoriteMovie(FavoriteMovie movie);

    @Query("SELECT * FROM favorite_movies WHERE id LIKE :id")
    LiveData<FavoriteMovie> findFavoriteMovie(int id);

    @Query("SELECT * FROM movies NATURAL JOIN favorite_movies")
    LiveData<Movie[]> loadAllFavoriteMovies();

    // Watch Later

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertWatchLaterMovie(WatchLaterMovie movie);

    @Delete
    void deleteWatchLaterMovie(WatchLaterMovie movie);

    @Query("SELECT * FROM watch_later_movies WHERE id LIKE :id")
    LiveData<WatchLaterMovie> findWatchLaterMovie(int id);

    @Query("SELECT * FROM movies NATURAL JOIN watch_later_movies")
    LiveData<Movie[]> loadAllWatchLaterMovies();

}

File: app/src/main/java/com/asmat/rolando/popularmovies/database/MoviesDAO.kt
package com.asmat.rolando.popularmovies.database

import androidx.lifecycle.LiveData
import androidx.room.*
import com.asmat.rolando.popularmovies.database.entities.FavoriteMovie
import com.asmat.rolando.popularmovies.database.entities.WatchLaterMovie
import io.reactivex.Completable
import io.reactivex.Single

@Dao
interface MoviesDAO {

    // Favorite

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertFavoriteMovie(movie: FavoriteMovie): Completable

    @Query("SELECT * FROM favorite_movies WHERE id LIKE :id")
    fun findFavoriteMovie(id: Int): LiveData<FavoriteMovie>

    @Query("DELETE FROM favorite_movies WHERE id LIKE :id ")
    fun deleteFavoriteMovie(id: Int): Single<Int>

    @Query("SELECT * FROM favorite_movies")
    fun loadAllFavoriteMovies(): LiveData<List<FavoriteMovie>>

    // Watch Later

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertWatchLaterMovie(movie: WatchLaterMovie): Completable

    @Query("SELECT * FROM watch_later_movies WHERE id LIKE :id")
    fun findWatchLaterMovie(id: Int): LiveData<WatchLaterMovie>

    @Query("DELETE FROM watch_later_movies WHERE id LIKE :id ")
    fun deleteWatchLaterMovie(id: Int): Single<Int>

    @Query("SELECT * FROM watch_later_movies")
    fun loadAllWatchLaterMovies(): LiveData<List<WatchLaterMovie>>

}